### PR TITLE
API: Cleaning numpy/__init__.py and main namespace - Part 2 [NEP 52]

### PR DESCRIPTION
**Pull request by mtsokol**
This PR is a second batch of NumPy main namespace changes.

First merged part available here: https://github.com/numpy/numpy/pull/24316.

## Scope of PR

- Removed `from ... import *` from the main namespace `numpy/__init__.py`, which is the main purpose of this PR. Changes listed below are direct consequences of this modification. A full main namespace cleanup will be a separate PR.
- Made submodules are imported lazily
- Removed a subset of names from the top namespace listed in the “remove” section in
https://github.com/numpy/numpy/issues/24306#issuecomment-1663564574. The current changelog:

    - `FLOATING_POINT_SUPPORT`
    - `FPE_DIVIDEBYZERO`
    - `FPE_INVALID`
    - `FPE_OVERFLOW`
    - `FPE_UNDERFLOW`
    - `NINF`, `NZERO`, `PINF`, `PZERO`
    - `UFUNC_BUFSIZE_DEFAULT`
    - `UFUNC_PYVALS_NAME`
    - `CLIP`, `WRAP`, `RAISE`
    - `BUFSIZE`
    - `ALLOW_THREADS`
    - `MAXDIMS`
    - `MAY_SHARE_EXACT`, `MAY_SHARE_BOUNDS`
    - `add_newdoc`, `add_docstring`, `add_newdoc_ufunc`
    
- Removed `issubdtype` from `lib` as it's already imported from `core` (top `__init__.py` had to remove one of them)
- Moved `add_newdoc`, `add_docstring`, `add_newdoc_ufunc` tests to `core`, as this module hosts their implementations
- Removed `Arrayterator` from `lib`'s `__all__`

### CODE REVIEW

File: numpy/__init__.py [hunk 1]
```diff
@@ -120,13 +120,90 @@
     __all__ = ['exceptions']
 
     from . import core
-    from .core import *
+    from .core import (
+        _no_nep50_warning, memmap,
+        iinfo, finfo,  # getlimits submodule exports
```
Comment (rgommers) [inline numpy/__init__.py:125]: no need for this comment I think?

File: numpy/__init__.py [hunk 2]
```diff
@@ -120,13 +120,90 @@
...
+        pi, positive, power, printoptions, prod, product, promote_types, 
+        ptp, put, putmask, rad2deg, radians, ravel, rec, recarray, reciprocal,
+        record, remainder, repeat, require, reshape, resize, result_type, 
+        right_shift, rint, roll, rollaxis, round, round_, sctype2char, 
+        sctypeDict, sctypes, searchsorted, set_printoptions,
+        set_string_function, setbufsize, seterr, seterrcall, shape,
+        shares_memory, short, sign, signbit, signedinteger, sin, single, 
+        singlecomplex, sinh, size, sometrue, sort, spacing, sqrt, square, 
+        squeeze, stack, std, str_, string_, subtract, sum, swapaxes, take,
+        tan, tanh, tensordot, test, timedelta64, trace, transpose,
```
Comment (rgommers) [inline numpy/__init__.py:169]: `test` can be left out here and in other places.

**Review by rgommers — COMMENTED** [review]
Getting rid of `import *` will be useful. The last commit looks close to ready after a rebase. Note that there are some more `__all__.remove` usages in `__init__.py` that can be removed (e.g. for `issubdtype`, `long`, `unicode`).

File: numpy/__init__.py (hunk 1 above)
Comment (mtsokol) [reply numpy/__init__.py]: Sure, I removed it! I think the list of imports will change - I will perform formatting and sorting in the end.

**Comment by mtsokol** [comment]
Hi @rgommers!

I pushed another batch of changes addressing review comments. Currently, in this PR:

- Removed `from ... import *` in the main namespace
- Made submodules being imported lazily 
- Removed a set of names from the top namespace listed in the “remove” section in https://github.com/numpy/numpy/issues/24306#issuecomment-1663564574:
    - `FLOATING_POINT_SUPPORT`
    - `FPE_DIVIDEBYZERO`
    - `FPE_INVALID`
    - `FPE_OVERFLOW`
    - `FPE_UNDERFLOW`
    - `NINF, NZERO, PINF, PZERO` - these are used quite a bit in the codebase, what should I use in place of `NZERO` and `PZERO`?
    - `UFUNC_BUFSIZE_DEFAULT`
    - `UFUNC_PYVALS_NAME`
    - `CLIP, WRAP, RAISE`
    - `BUFSIZE` - this is used in a few ufunc and cleanup tests. I would import it from core directly then.
    - `ALLOW_THREADS`
    - `MAXDIMS`

**Comment by ngoldbaum** [comment]
> I've got one question: These are used quite a bit in the codebase, what should I use in place of NZERO and PZERO? Just 0.0 and -0.0?

I'd probably change the usages in the tests to import these constants from `numpy.core.umath`. They probably don't need explicit docs anymore now that they're not public, so you could remove the entries for them from `add_newdoc.py`.

**Comment by ngoldbaum** [comment]
Great to see `import *` starting to go away! :rocket:

**Comment by mtsokol** [comment]
Hi @rgommers @ngoldbaum,

This PR is ready for a review - some CI staged failed due to the fact that Pandas uses `np.NINF`, which was removed here. As before, I will propagate these changes in separate PRs in other related libraries. 

This work's primary purpose is to remove `*` imports and clean a few names that are surely being removed from the top namespace. Next stage of cleanup will be in a separate PR.

**Comment by ngoldbaum** [comment]
Tests are failing because pandas fails to import in the doctests build. Will be fixed once https://github.com/pandas-dev/pandas/pull/54468/files is merged and a new pandas wheel gets uploaded.

File: numpy/__init__.py [hunk 3]
```diff
@@ -116,26 +116,107 @@
...
+    for sa in sized_aliases:
+        try:
+            globals()[sa] = getattr(core, sa)
+        except AttributeError:
+            pass
+    del sa, sized_aliases
+
     from . import lib
     # NOTE: to be revisited following future namespace cleanup.
     # See gh-14454 and gh-15672 for discussion.
```
Comment (ngoldbaum) [inline numpy/__init__.py:186]: I believe this PR is the namespace cleanup this note references, so best to delete it.

**Review by ngoldbaum — APPROVED** [review]
Just one comment that should be deleted I think, otherwise LGTM!

File: numpy/__init__.py (hunk 3 above)
Comment (mtsokol) [reply numpy/__init__.py]: Sure! Removed

**Comment by rgommers** [comment]
> Tests are failing because pandas fails to import in the doctests build. Will be fixed once https://github.com/pandas-dev/pandas/pull/54468/files is merged and a new pandas wheel gets uploaded.

Looks like we're in luck; pandas uploads nightlies frequently and it's already fixed, the job passed.

File: numpy/doc/constants.py [hunk 4]
```diff
@@ -175,109 +175,6 @@ def add_newdoc(module, name, doc):
 
     """)
```
Comment (rgommers) [inline numpy/doc/constants.py:177]: I think `numpy/doc` can be removed completely as well - anything that looks useful still and isn't already present can be moved to the html docs (it can't be much), and then it's `rm -rf` time I think. It shouldn't be used in any upstream library I think.

**Review by rgommers — CHANGES_REQUESTED** [review]
@mtsokol this is looking good but needs a few tweaks I think. I checked the diff between `np.__dir__()` in main vs this PR, and this is the list I get:

```
{'ALLOW_THREADS',
 'BUFSIZE',
 'CLIP',
 'FLOATING_POINT_SUPPORT',
 'FPE_DIVIDEBYZERO',
 'FPE_INVALID',
 'FPE_OVERFLOW',
 'FPE_UNDERFLOW',
 'MAXDIMS',
 'MAY_SHARE_BOUNDS',
 'MAY_SHARE_EXACT',
 'NAN',
 'NINF',
 'NZERO',
 'PINF',
 'PZERO',
 'RAISE',
 'UFUNC_BUFSIZE_DEFAULT',
 'UFUNC_PYVALS_NAME',
 'WRAP',
 '_UFUNC_API',
 '_add_newdoc_ufunc',
 'add_docstring',
 'add_newdoc',
 'add_newdoc_ufunc',
 'ctypeslib',
 'fft',
 'ma',
 'polynomial',
 'random'}
```

The last 5 items are unintentional removals it looks like; they're lazily loaded with `__getattr__` now so should be added to `__dir__()`.

```
Then, I think you also would like to test the consistency between `np.__dir__()` and `np.__all__` probably. Right now I see:
```python
>>> {s for s in set(np.__dir__()) - set(np.__all__) if not s.startswith('_')}
{'version', 'dtypes', 'defmatrix', 'test', 'testing', 'abs', 'lib'}
```

`defmatrix` and `test` will go away, and `version` should be removed later on, but `dtypes`, `testing`, `lib` and `abs` should be in both I'd think. May be good to add a test that checks that they contain the same, modulo items starting with `_`?

File: numpy/__init__.py [hunk 5]
```diff
@@ -116,26 +116,105 @@
...
+    )
+
+    sized_aliases = ["int8", "int16", "int32", "int64", "intp", 
+                     "uint8", "uint16", "uint32", "uint64", "uintp",
+                     "float16", "float32", "float64", "float96", "float128",
+                     "complex64", "complex128", "complex192", "complex256"]
+    
+    for sa in sized_aliases:
+        try:
+            globals()[sa] = getattr(core, sa)
```
[the remark this answers is not in the document]
Comment (mtsokol) [reply numpy/__init__.py]: Sure! I removed that list to only these four tentative aliases.

File: numpy/doc/constants.py (hunk 4 above)
Comment (mtsokol) [reply numpy/doc/constants.py]: In the `numpy/doc`, only two files were left: `constants.py` and `ufuncs.py`. I adjusted and moved the first one to the `constants.rst` file. I left the second one as, according to https://github.com/numpy/numpy/pull/17222, it still needs to be properly merged with `doc/source/reference/ufuncs.rst`.

**Comment by mtsokol** [comment]
@rgommers I adjusted `__all__` and `__dir__` so they yield the same content (I also check it in a separate test in `test_public_api.py`). 
A comment: This new test showed that in testing environment a bunch of redundant names end up in `np.__dir__`, such as `conftest`, `tests`, due to the fact that `numpy.__init__.py` is called after `globals` was augmented with new names. 
I wonder: Do we want to keep it that way and manually remove these items as now? Or do we want to move to statically defined list for `__dir__()`? E.g. use `__all__` there, and have a test that make sure that `globals` is not growing with more unexpected content. WDYT?

**Comment by rgommers** [comment]
I'm not sure about the use of `globals()`, it isn't great indeed. But it's a little tricky to figure out what to replace it with. We can punt on that for now I'd say.

The one thing I did notice is that the list of public symbols still isn't 100% correct for submodule names. From https://numpy.org/neps/nep-0052-python-api-cleanup.html#cleaning-up-the-submodule-structure I think `__dir__` and `__all__` should contain all the namespaces that are listed as "regular", "special-purpose" and "legacy", and the ones listed for "removal" can be left out (or included now and removed later). And all the other ones can be lazy-loaded.

File: numpy/__init__.py [hunk 6]
```diff
@@ -226,16 +320,21 @@ def __getattr__(attr):
...
+        # TODO: move away from using `globals` to a statically defined 
+        # list. With `globals`, when running in a testing context 
+        # a bunch of random names fall into global scope, such as
+        # `conftest` or `distutils`.
+        public_symbols = (
+            globals().keys() | __numpy_submodules__
+        )
+        public_symbols -= {
+            "core", "matrixlib", "matlib", "test", "tests", "conftest", 
+            "version", "typing", "f2py", "compat", "array_api", "distutils"
```
Comment (rgommers) [inline numpy/__init__.py:336]: Some of these are public - `test`, `distutils`, `typing` and `array_api` at least, and those will stay for 2.0.  The others can be left out.

**Comment by mtsokol** [comment]
> The one thing I did notice is that the list of public symbols still isn't 100% correct for submodule names. From https://numpy.org/neps/nep-0052-python-api-cleanup.html#cleaning-up-the-submodule-structure I think `__dir__` and `__all__` should contain all the namespaces that are listed as "regular", "special-purpose" and "legacy", and the ones listed for "removal" can be left out (or included now and removed later). And all the other ones can be lazy-loaded.

In this PR I wanted to mostly focus on removing `from ... import *`. The PR with Part 3 that I'm working on I cover exact refactor of main namespace (following NEP), but I can include all target submodules here already. Should I move it here?

File: numpy/__init__.py (hunk 6 above)
Comment (mtsokol) [reply numpy/__init__.py]: Sure! I updated main namespace, so it matches the list from NEP 52.

**Comment by rgommers** [comment]
This looks good to me now:

```python
>>> l = np.__dir__()
>>> l2 = [str(s) for s in np.load('../bldnumpy/main.npy')]  # __dir__ from current `main`
>>> set(l2) - set(l)
{'PZERO', 'add_newdoc', 'MAY_SHARE_EXACT', 'ALLOW_THREADS', 'UFUNC_BUFSIZE_DEFAULT', 'CLIP', 'RAISE', 'NAN', 'UFUNC_PYVALS_NAME', 'NINF', 'FPE_OVERFLOW', 'version', 'FPE_INVALID', 'FPE_UNDERFLOW', '_UFUNC_API', 'MAXDIMS', 'WRAP', 'MAY_SHARE_BOUNDS', 'add_docstring', 'NZERO', 'add_newdoc_ufunc', 'FPE_DIVIDEBYZERO', 'PINF', '_add_newdoc_ufunc', 'FLOATING_POINT_SUPPORT', 'BUFSIZE'}
>>> set(l) - set(l2)
{'f2py', 'array_api', 'typing', 'distutils', '__numpy_submodules__'}

>>> {s for s in set(np.__dir__()) - set(np.__all__)}
{'__getattr__', '__package__', '_NoValue', '_pyinstaller_hooks_dir', '__builtins__', '__spec__', '_typing', '__loader__', '__git_version__', '__config__', '__all__', '_globals', '__path__', '__doc__', '__NUMPY_SETUP__', '__dir__', '_distributor_init', '__file__', '_pytesttester', '_version', '__future_scalars__', '__cached__', '_utils', '_type_info', '_int_extended_msg', '__name__', '__former_attrs__', '_CopyMode', '_msg', '__numpy_submodules__', '_mat', '_specific_msg', '_version_meson'}

>>> {s for s in set(np.__dir__()) - set(np.__all__) if not s.startswith('_')}
set()
>>> {s for s in set(np.__all__) - set(np.__dir__()) if not s.startswith('_')}
set()
```

File: numpy/lib/arrayterator.py [hunk 7]
```diff
@@ -66,7 +66,7 @@ class Arrayterator:
     Examples
     --------
     >>> a = np.arange(3 * 4 * 5 * 6).reshape(3, 4, 5, 6)
-    >>> a_itor = np.lib.Arrayterator(a, 2)
+    >>> a_itor = np.lib.arrayterator.Arrayterator(a, 2)
```
Comment (rgommers) [inline numpy/lib/arrayterator.py:69]: `np.lib.Arrayterator` is still present given the current state of this PR. Was this change needed to make the doc build happy? If not, may be better to revert it.

**Review by rgommers — APPROVED** [review]
This looks good to me now. The added test to ensure that `__dir__()` and `__all__` contain the same public members is nice and will be useful going forward.

One last question about `Arrayterator`, since that's a change to `numpy.lib` that would have been nicer to avoid.

All the removed items from the main namespace here seem uncontroversial, and you've checked that none of them are used in Pandas, scikit-learn, SciPy, Matplotlib, JAX, CuPy et al. anymore (or, in most cases, never were used to begin with).

**Comment by mtsokol** [comment]
@rgommers One last thing - I double checked other libraries that are impacted and I noticed that SciPy uses ```from numpy.lib import add_newdoc```, and I removed `add_newdoc` also from `lib` here, but shouldn't - let me revert that change.

File: numpy/lib/arrayterator.py (hunk 7 above)
Comment (mtsokol) [reply numpy/lib/arrayterator.py]: I did it in advance, but `lib` refactor is not a part of this PR - reverted.

**Comment by rgommers** [comment]
Thanks Mateusz! I know this is fiddly, but in the long run it'll be worth it to have PRs that are independent and scoped to a well-defined purpose.

I'll look at merging this once CI is happy.

**Comment by mtsokol** [comment]
> Thanks Mateusz! I know this is fiddly, but in the long run it'll be worth it to have PRs that are independent and scoped to a well-defined purpose.
> 
> I'll look at merging this once CI is happy.

@rgommers Looks that CI completed. I did one more round checking usages of removed items in other libraries and found one in SciPy that I missed: https://github.com/scipy/scipy/pull/19045

**Review by rgommers — APPROVED** [review]
Thanks @mtsokol, all LGTM now. I ran out of time with the many other NumPy changes; will leave this open right now just in case someone else wants to have a look. I'll plan to merge it tomorrow.

**Comment by rgommers** [comment]
In it went - thanks again Mateusz!
